src/Facility.Definition/ServiceInfo.cs

```diff
@@ -30,10 +31,19 @@ public ServiceInfo(string name, IEnumerable<ServiceMemberInfo> members, IEnumera
 			foreach (var fieldGroup in GetDescendants().OfType<ServiceFieldInfo>().GroupBy(x => x.TypeName))
 			{
 				var type = ServiceTypeInfo.TryParse(fieldGroup.Key, FindMember);
-				if (type != null)
-					m_typesByName.Add(fieldGroup.Key, type);
-				else
+				if (type == null)
+				{
 					AddValidationErrors(fieldGroup.Select(x => new ServiceDefinitionError($"Unknown field type '{x.TypeName}'.", x.GetPart(ServicePartKind.TypeName)?.Position)));
+					return;
+				}
+
+				m_typesByName.Add(fieldGroup.Key, type);
+
+				// Ensuring correct usage of [validate] requires knowledge of type relationships, so it must be done here
+				foreach (var field in fieldGroup.Where(x => x.Attributes.Any(a => a.Name == "validate")))
```

Comment:
The `Where` doesn't seem necessary since `EnsureProperValidateUsage` does nothing if there are none.